worlds/lingo/rules.py

```diff
@@ -0,0 +1,98 @@
+        if door is None:
+            return True
+        else:
+            if self._lingo_can_open_door(room, room if door.room is None else door.room, door.door, world,
+                                         player, player_logic):
+                return True
+            return False
+
+    def lingo_can_use_pilgrimage(self, player: int, player_logic: LingoPlayerLogic):
+        fake_pilgrimage = [
+            ["Starting Room", "Main Door"], ["Second Room", "Exit Door"],
+            ["Crossroads", "Tower Entrance"], ["Orange Tower Fourth Floor", "Hot Crusts Door"],
+            ["Outside The Initiated", "Shortcut to Hub Room"], ["Orange Tower First Floor", "Shortcut to Hub Room"],
+            ["Directional Gallery", "Shortcut to The Undeterred"], ["Orange Tower First Floor", "Salt Pepper Door"],
+            ["The Tenacious", "Shortcut to Hub Room"], ["Outside The Agreeable", "Tenacious Entrance"]
+        ]
+        viable_option = True
+        for entrance in fake_pilgrimage:
+            if not self.has(player_logic.ITEM_BY_DOOR[entrance[0]][entrance[1]], player):
+                viable_option = False
+                break
+        return viable_option
+
+    def lingo_can_use_location(self, location: PlayerLocation, room_name: str, world: MultiWorld, player: int,
+                               player_logic: LingoPlayerLogic):
+        for panel in location.panels:
+            panel_room = room_name if panel.room is None else panel.room
+            if not self._lingo_can_solve_panel(room_name, panel_room, panel.panel, world, player, player_logic):
+                return False
+        return True
+
+    def lingo_can_use_mastery_location(self, world: MultiWorld, player: int):
+        return self.has("Mastery Achievement", player, get_option_value(world, player, "mastery_achievements"))
+
+    def _lingo_can_open_door(self, start_room: str, room: str, door: str, world, player,
+                             player_logic: LingoPlayerLogic):
+        """
+        Determines whether a door can be opened
+        """
+        item_name = player_logic.ITEM_BY_DOOR[room][door]
+        if item_name in StaticLingoLogic.PROGRESSIVE_ITEMS:
+            progression = StaticLingoLogic.PROGRESSION_BY_ROOM[room][door]
+            return self.has(item_name, player, progression.index)
+        else:
+            return self.has(item_name, player)
+
+    def _lingo_can_solve_panel(self, start_room: str, room: str, panel: str, world, player,
+                               player_logic: LingoPlayerLogic):
+        """
+        Determines whether a panel can be solved
+        """
+        if start_room != room and not self.can_reach(room, "Region", player):
+            return False
+        panel_object = StaticLingoLogic.PANELS_BY_ROOM[room][panel]
+        for req_room in panel_object.required_rooms:
+            if not self.can_reach(req_room, "Region", player):
+                return False
```

Comment:
performing can_reach on a region inside a can_reach / access condition for a location is hella spooky.
When a new region becomes accessible, but nothing else new happens, the collection state does not become "stale", meaning locations aren't checked again for reachability. This can lead to seeds randomly failing to generate.

This PR addresses that:
https://github.com/ArchipelagoMW/Archipelago/pull/1764

Witness solves this via the use of event items.

Reply:
Converted to event items, PTAL

Reply:
Does this only check for region event items that aren't already the location's parent region anyway? If so, you can mark this as resolved.

Reply:
It checks that the panel's region is accessible, but only if the panel is in a different region than the location that we're determining access for.